security/src/main/java/io/micronaut/security/authentication/BlockingAuthenticationProviderFactory.java

```diff
@@ -0,0 +1,79 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.security.authentication;
+
+import io.micronaut.context.annotation.EachBean;
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.scheduling.TaskExecutors;
+import jakarta.inject.Named;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+
+import java.util.concurrent.ExecutorService;
+
+/**
+ * A factory for adapting {@link BlockingAuthenticationProvider} beans to expose them as {@link AuthenticationProvider}.
+ *
+ * @since 4.5.0
+ */
+@Factory
+@Internal
+class BlockingAuthenticationProviderFactory {
+
+    private final Scheduler scheduler;
+
+    BlockingAuthenticationProviderFactory(@Named(TaskExecutors.BLOCKING) ExecutorService executorService) {
+        this.scheduler = Schedulers.fromExecutorService(executorService);
+    }
+
+    /**
+     * Creates an adapted {@link AuthenticationProvider} for each provided instance of {@link BlockingAuthenticationProvider}.
+     *
+     * <p>
+     * NOTE - If there are multiple instances of {@link BlockingAuthenticationProvider} in the application context, then they
+     * must be annotated with a {@link jakarta.inject.Qualifier} such as {@link Named}.
+     * </p>
+     *
+     * @param blockingAuthenticationProvider An instance of {@link BlockingAuthenticationProvider} to be adapted
+     * @return An {@link AuthenticationProvider} adapted from the blocking provider
+     * @param <T> The request type
+     */
+    @EachBean(BlockingAuthenticationProvider.class)
+    <T> AuthenticationProvider<T> createAuthenticationProvider(BlockingAuthenticationProvider<T> blockingAuthenticationProvider) {
+        return new BlockingAuthenticationProviderAdapter<>(blockingAuthenticationProvider, scheduler);
+    }
+
+    private static final class BlockingAuthenticationProviderAdapter<T> implements AuthenticationProvider<T> {
+
+        private final BlockingAuthenticationProvider<T> blockingAuthenticationProvider;
+
+        private final Scheduler scheduler;
+
+        private BlockingAuthenticationProviderAdapter(BlockingAuthenticationProvider<T> blockingAuthenticationProvider, Scheduler scheduler) {
+            this.blockingAuthenticationProvider = blockingAuthenticationProvider;
+            this.scheduler = scheduler;
+        }
+
+        @Override
+        public Publisher<AuthenticationResponse> authenticate(T httpRequest, AuthenticationRequest<?, ?> authenticationRequest) {
+            return Mono.fromCallable(() -> blockingAuthenticationProvider.authenticate(httpRequest, authenticationRequest))
```

Comment:
I would expect to use the new API directly in the filter, not have it wrapped in a reactive

Reply:
Preferable this change should allow not engage any reactive filtering when all the filters are non-reactive

Reply:
I think that should be attempted in the next major version of Micronaut Security.

Reply:
I have removed the factory and I only create `ReactiveAuthenticationProvider` if users define any or if the user authentication provider is blocking https://github.com/micronaut-projects/micronaut-security/pull/1526/commits/c795eaada8c7253f0a82eef4334195a575273eb8